[Expanded Weights] add ability to not specify batch size

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80944
* #80943
* #80895
* #80892

Opacus has been asking for the ability to not specify a batch size. Previously a user had to do
`call_for_per_sample_grads(module, batch_size)(*args, **kwargs)`
They rightfully pointed out that in most cases when you're passing a single argument to a module's forward function, it seems repetitive to specify the batch_size. The argument here is that in cases where a user was passing more than one argument, we might not know what the batch size is if they don't match. So, this lets a user not specify a batch size (or pass it as None), meaning that
`call_for_per_sample_grad(linear_module)(torch.randn(5, 4))`
 now works and has a batch size of 5

If there are multiple tensor arguments with different batch sizes, we fail, even if one of the inputs wouldn't have been used by the module because we can't tell which batch size we should be using.

Commit:
[Expanded Weights] add ability to not specify batch size

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] add ability to not specify batch size"




[ghstack-poisoned]

zou3519 commented on torch/nn/utils/_per_sample_grad.py:
```diff
@@ -38,20 +43,33 @@ def call_for_per_sample_grads(module, batch_size, loss_reduction="mean"):
             return og_tensor
 
+    def compute_batch_size(*args, **kwargs):
+        args_and_kwargs = tree_flatten(args)[0] + tree_flatten(kwargs.values())[0]
```
This imposes a new restriction that args and kwargs contains pytree-able data types. For function transforms (vmap) we assume that and that's fine, but in general this will not work with arbitrary nn.Modules.

But tbh this should work in most cases

samdow replied:
Hmmm fair. If it's not pytree-able, it will still work if the user explicitly gives a batch size because it won't call this function. I'm not seeing an `is_pytreeable` function but if I put this in a try/catch, I should be able to give the user an error message to that effect

zou3519 commented on torch/nn/utils/_per_sample_grad.py:
```diff
@@ -38,20 +43,33 @@ def call_for_per_sample_grads(module, batch_size, loss_reduction="mean"):
+    def compute_batch_size(*args, **kwargs):
+        args_and_kwargs = tree_flatten(args)[0] + tree_flatten(kwargs.values())[0]
+        batch_size = None
+        for arg in args_and_kwargs:
+            if isinstance(arg, torch.Tensor):
+                arg_batch_size = arg.shape[0]  # we assume batch size is the first dim
+                if batch_size is not None and batch_size != arg_batch_size:
+                    raise RuntimeError("When computing batch size, found at least one input with batch size "
+                                       f"{batch_size} and one with batch size {arg_batch_size}. Please specify it "
+                                       "explicitly using the batch size kwarg in call_for_per_sample_grads")
+                batch_size = arg_batch_size
```
nit: can save on indentation with continues:
```
for arg in args_and_kwargs:
  if not isinstance(arg, torch.Tensor):
    continue
  if batch_size is not None and batch_size != arg_batch_size:
    raise ...
```

zou3519 reviewed (approved):
LGTM, we are making some assumptions on what args and kwargs look like though

Commit:
Update on "[Expanded Weights] add ability to not specify batch size"


Opacus has been asking for the ability to not specify a batch size. Previously a user had to do
`call_for_per_sample_grads(module, batch_size)(*args, **kwargs)`
They rightfully pointed out that in most cases when you're passing a single argument to a module's forward function, it seems repetitive to specify the batch_size. The argument here is that in cases where a user was passing more than one argument, we might not know what the batch size is if they don't match. So, this lets a user not specify a batch size (or pass it as None), meaning that
`call_for_per_sample_grad(linear_module)(torch.randn(5, 4))`
 now works and has a batch size of 5

If there are multiple tensor arguments with different batch sizes, we fail, even if one of the inputs wouldn't have been used by the module because we can't tell which batch size we should be using.

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] add ability to not specify batch size"


Opacus has been asking for the ability to not specify a batch size. Previously a user had to do
`call_for_per_sample_grads(module, batch_size)(*args, **kwargs)`
They rightfully pointed out that in most cases when you're passing a single argument to a module's forward function, it seems repetitive to specify the batch_size. The argument here is that in cases where a user was passing more than one argument, we might not know what the batch size is if they don't match. So, this lets a user not specify a batch size (or pass it as None), meaning that
`call_for_per_sample_grad(linear_module)(torch.randn(5, 4))`
 now works and has a batch size of 5

If there are multiple tensor arguments with different batch sizes, we fail, even if one of the inputs wouldn't have been used by the module because we can't tell which batch size we should be using.

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] add ability to not specify batch size"


Opacus has been asking for the ability to not specify a batch size. Previously a user had to do
`call_for_per_sample_grads(module, batch_size)(*args, **kwargs)`
They rightfully pointed out that in most cases when you're passing a single argument to a module's forward function, it seems repetitive to specify the batch_size. The argument here is that in cases where a user was passing more than one argument, we might not know what the batch size is if they don't match. So, this lets a user not specify a batch size (or pass it as None), meaning that
`call_for_per_sample_grad(linear_module)(torch.randn(5, 4))`
 now works and has a batch size of 5

If there are multiple tensor arguments with different batch sizes, we fail, even if one of the inputs wouldn't have been used by the module because we can't tell which batch size we should be using.

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] add ability to not specify batch size"


Opacus has been asking for the ability to not specify a batch size. Previously a user had to do
`call_for_per_sample_grads(module, batch_size)(*args, **kwargs)`
They rightfully pointed out that in most cases when you're passing a single argument to a module's forward function, it seems repetitive to specify the batch_size. The argument here is that in cases where a user was passing more than one argument, we might not know what the batch size is if they don't match. So, this lets a user not specify a batch size (or pass it as None), meaning that
`call_for_per_sample_grad(linear_module)(torch.randn(5, 4))`
 now works and has a batch size of 5

If there are multiple tensor arguments with different batch sizes, we fail, even if one of the inputs wouldn't have been used by the module because we can't tell which batch size we should be using.

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] add ability to not specify batch size"


Opacus has been asking for the ability to not specify a batch size. Previously a user had to do
`call_for_per_sample_grads(module, batch_size)(*args, **kwargs)`
They rightfully pointed out that in most cases when you're passing a single argument to a module's forward function, it seems repetitive to specify the batch_size. The argument here is that in cases where a user was passing more than one argument, we might not know what the batch size is if they don't match. So, this lets a user not specify a batch size (or pass it as None), meaning that
`call_for_per_sample_grad(linear_module)(torch.randn(5, 4))`
 now works and has a batch size of 5

If there are multiple tensor arguments with different batch sizes, we fail, even if one of the inputs wouldn't have been used by the module because we can't tell which batch size we should be using.

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] add ability to not specify batch size"


Opacus has been asking for the ability to not specify a batch size. Previously a user had to do
`call_for_per_sample_grads(module, batch_size)(*args, **kwargs)`
They rightfully pointed out that in most cases when you're passing a single argument to a module's forward function, it seems repetitive to specify the batch_size. The argument here is that in cases where a user was passing more than one argument, we might not know what the batch size is if they don't match. So, this lets a user not specify a batch size (or pass it as None), meaning that
`call_for_per_sample_grad(linear_module)(torch.randn(5, 4))`
 now works and has a batch size of 5

If there are multiple tensor arguments with different batch sizes, we fail, even if one of the inputs wouldn't have been used by the module because we can't tell which batch size we should be using.

[ghstack-poisoned]

Commit:
Update on "[Expanded Weights] add ability to not specify batch size"


Opacus has been asking for the ability to not specify a batch size. Previously a user had to do
`call_for_per_sample_grads(module, batch_size)(*args, **kwargs)`
They rightfully pointed out that in most cases when you're passing a single argument to a module's forward function, it seems repetitive to specify the batch_size. The argument here is that in cases where a user was passing more than one argument, we might not know what the batch size is if they don't match. So, this lets a user not specify a batch size (or pass it as None), meaning that
`call_for_per_sample_grad(linear_module)(torch.randn(5, 4))`
 now works and has a batch size of 5

If there are multiple tensor arguments with different batch sizes, we fail, even if one of the inputs wouldn't have been used by the module because we can't tell which batch size we should be using.

[ghstack-poisoned]

samdow commented:
@pytorchbot merge -g
